### DISCUSSION TITLE
Building a pyodide/python wrapper around a c library using autotools

### DISCUSSION THREAD

**Discussion by NelisDrost** [pyodide/pyodide#4456] [Q&A]
I have a c library that I want to make available via pyodide in WASM, which I believe takes the following steps:

1. Create python bindings for that c library (in my case using ctypes)
2. Create a setup.py that builds both the underlying library and python bindings into a wheel
3. `pyodide build` to create a wasm wheel

I'm fairly comfortable with step 1, but use some guidance on step 2. Specifically, the c library I'm using builds with autotools (i.e.: it has a `configure` script that creates Makefiles, which can then be run with `make`). The examples I've seen for pyodide so far are primarily in C++ and use CMake.

Is it possible to `pyodide build` using the current build system? If so, how do I represent that in setup.py?
If that's not possible, is the solution to replace that build system with CMake - will that work?

The c library, for reference: https://sourceforge.net/projects/libassp/

**Comment by NelisDrost** [comment]
A follow up question, that I probably should have started with - are there any gotchas that might prevent this working, e.g.:
- I'm targeting a C, rather than C++ library
- The library uses stdio::FILE
- The library makes heavy use of void* pointers to arbitrary chunks of data

**Comment by hoodmane** [comment]
We build plenty of C libraries built with configure and make. If you don't use threads, networking, gui libraries, signal handling or stuff like that, the most common issues are around the expectation that pointers are 32 bits.

To start with you should check that your library can be built with Emscripten. For reference, libxslt is a C library build with autotools and Make:
https://github.com/pyodide/pyodide/blob/main/packages/libxslt/meta.yaml

I think the best way would be to include your C library as a `shared_library`. Then you can make a pure Python module in ctypes that depends on it. But I would recommend making a binary extension using CFFI instead. This will run faster and be less error prone.
https://cffi.readthedocs.io/en/stable/

**Reply by NelisDrost** [reply]
Excellent - many thanks for the pointers!

I've successfully built using emscripten, using the commands below. That results in .wasm files for many of the components, but not the overall library (I do get .a, .la, .lai files though, where normally I would use the .so). That may be normal/expected behavior though?
```sh
emconfigure ./configure --disable-shared
emmake make
```
(omitting --disable-shared results in duplicate symbol errors)

I should be able to make a bit more progress from here, thanks for getting me started!

**Reply by hoodmane** [reply]
Yeah that sounds like you successfully built a static library. So you could make a static library recipe by doing that and adding `cp mylib.a ${DISTDIR}` to the end. Then follow the instructions here to make a `setup.py` file for CFFI with your static library:
https://cffi.readthedocs.io/en/stable/overview.html#main-mode-of-usage

It's probably easier if you get the cffi module working against a native linux build of your library first.
